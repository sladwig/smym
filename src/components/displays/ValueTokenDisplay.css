.value-token {
    transition: color 120ms ease-in-out ;
}
