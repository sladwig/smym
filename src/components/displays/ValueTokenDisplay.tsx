import React, { CSSProperties, useCallback } from 'react';
import { isChar, CharCharacterT, AnyCharacterT } from './types';
import { useTransition, animated } from 'react-spring';
import { display } from './display';

const transitionOptions = {
    initial: { transform: 'translateY(-40px)', opacity: 0 },
    enter: { transform: 'translateY(0)', opacity: 1 },
    trail: 24,
    duration: 50,
    unique: true,
    expires: 0,
    key: (item: { key: number }) => `fixed-${item.key}`,
};

export const ValueTokenDisplay = ({
    word,
    style,
}: {
    word: { characters: AnyCharacterT[]; value: string };
    style?: CSSProperties;
}) => {
    const onlyChars = word.characters.filter(isChar);

    let firstPosition = onlyChars.length > 0 ? onlyChars[0].position : 0;
    let lastPosition = onlyChars.length > 0 ? onlyChars[onlyChars.length - 1].position : 0;

    const fixed: Array<CharCharacterT & { key: number }> = [
        { type: 'char', value: '.', position: lastPosition++, key: 0 },
        { type: 'char', value: '0', position: lastPosition++, key: 1 },
        { type: 'char', value: '0', position: lastPosition++, key: 2 },
    ];

    const commaSplittedValues = word.value.split(/,|\./);
    const fixRestPosition = 1 < commaSplittedValues.length ? commaSplittedValues[1].length + 1 : 0;

    const transitionEuro = useTransition(
        { type: 'char', value: 'â‚¬', position: firstPosition, key: 4 } as CharCharacterT & {
            key: number;
        },
        transitionOptions,
    );

    const transition = useTransition(fixed.slice(fixRestPosition), {
        ...transitionOptions,
        enter: { transform: 'translateY(0)', opacity: 0.6 },
    });
    const toAnimatedDiv = useCallback(
        (style: {}, item: CharCharacterT & { key: number }) => (
            <animated.div key={`fix-${item.key}`} style={style}>
                {display(item)}
            </animated.div>
        ),
        [],
    );

    return (
        <div className="row-flex value-token" style={style}>
            {transitionEuro(toAnimatedDiv)}
            {word.characters.map(display)}
            {transition(toAnimatedDiv)}
        </div>
    );
};
