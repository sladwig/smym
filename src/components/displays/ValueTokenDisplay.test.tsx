import { render } from '@testing-library/react';
import React from 'react';
import { ValueTokenDisplay } from './ValueTokenDisplay';

describe('ValueTokenDisplay', () => {
    test('renders', () => {
        const { container } = render(<ValueTokenDisplay word={{ characters: [], value: '' }} />);
        expect(container).toMatchSnapshot();
    });
});
