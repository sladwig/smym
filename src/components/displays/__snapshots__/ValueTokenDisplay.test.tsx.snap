// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueTokenDisplay renders 1`] = `
<div>
  <div
    class="row-flex value-token"
  >
    <div
      style="transform: translateY(-40px); opacity: 0;"
    >
      <span
        class="position-emitter"
      >
        â‚¬
        <span
          class="position-area-overlay"
        >
          <span
            class="position-area"
          />
          <span
            class="position-area"
          />
        </span>
      </span>
    </div>
    <div
      style="transform: translateY(-40px); opacity: 0;"
    >
      <span
        class="position-emitter"
      >
        .
        <span
          class="position-area-overlay"
        >
          <span
            class="position-area"
          />
          <span
            class="position-area"
          />
        </span>
      </span>
    </div>
    <div
      style="transform: translateY(-40px); opacity: 0;"
    >
      <span
        class="position-emitter"
      >
        0
        <span
          class="position-area-overlay"
        >
          <span
            class="position-area"
          />
          <span
            class="position-area"
          />
        </span>
      </span>
    </div>
    <div
      style="transform: translateY(-40px); opacity: 0;"
    >
      <span
        class="position-emitter"
      >
        0
        <span
          class="position-area-overlay"
        >
          <span
            class="position-area"
          />
          <span
            class="position-area"
          />
        </span>
      </span>
    </div>
  </div>
</div>
`;
